docs/cryptodoc/src/05_09_frodokem.rst

```diff
@@ -0,0 +1,298 @@
+the correct data is used at each step of the computation.
+More concretely, strong types are used for the different kinds of seeds, random samples, matrix
+representations, and plaintexts as well as for the values
+:math:`salt`, :math:`k`, and :math:`pkh` as found in [FrodoKEM-ISO]_.
+
+..  _pubkey/frodokem/modes:
+
+Modes and Constants
+^^^^^^^^^^^^^^^^^^^
+
+The implementation realizes all parameter sets shown in Table
+:ref:`Supported FrodoKEM parameters <pubkey/frodokem/parameter_table>`.
+Parameter sets are represented as instances of the enum-like class ``FrodoKEMMode``.
+When a mode has been selected, the parameters as well as the function :math:`\text{SHAKE}` of [FrodoKEM-ISO]_ are instantiated
+via the ``FrodoKEMConstants`` class. Furthermore, this class contains
+the distribution table entries required for sampling from the error distribution (see Table A.4 of [FrodoKEM-ISO]_).
+
+..  _pubkey/frodokem/matrix_operations:
+
+Matrix Operations
+^^^^^^^^^^^^^^^^^
+
+FrodoKEM heavily relies on matrix operations which are implemented in Botan
+within the ``FrodoMatrix`` class. Specifically, factory methods to create or store matrices
+according to the :math:`Encode` (Section 7.2 of [FrodoKEM-ISO]_),
+:math:`Decode` (Section 7.2), :math:`Pack` (Section 7.3), :math:`Unpack` (Section 7.3),
+and :math:`SampleMatrix` (Section 7.4 and 7.5) functions are provided, thereby realizing
+the respective functions in accordance with the specification.
+
+The implementation does not explicitly provide the :math:`Gen` function (Section 7.6
+of [FrodoKEM-ISO]_) to create the large public matrix :math:`A` from :math:`seed_A`.
+Instead, to avoid having the entire matrix :math:`A` in memory for just one use per operation,
+the desired elements of :math:`A` are derived from :math:`seed_A` on demand when matrix
+multiplication is performed. This is done via callable generator functions which generate
+the required rows of :math:`A` and correspond to
+Section 7.6.1 or 7.6.2 of [FrodoKEM-ISO]_, depending on whether AES-128 or SHAKE-128 is used.
+
+Moreover, the implementation does not perform the transpose operation on the matrix
+:math:`S^T` to obtain :math:`S` as in the pseudocode of Section 8 of [FrodoKEM-ISO]_
+when computing :math:`B = AS + E` and :math:`B'S` during key generation and decapsulation, respectively. Instead, it performs
+the matrix operations directly on input :math:`S^T` to obtain the desired result
+without an expensive transpose operation.
+
+Consequently, each unique combination of matrix operations used by FrodoKEM
+:ref:`Key Generation <pubkey/frodokem/key_generation>`,
+:ref:`Encapsulation <pubkey/frodokem/encapsulation>` and
+:ref:`Decapsulation <pubkey/frodokem/decapsulation>` is implemented as a
+``FrodoMatrix`` factory method. Hence, the operations :math:`AS + E`, :math:`S'A + E'`,
+:math:`S'B + E''` and :math:`B'S` each have a corresponding method. Additionally, generic
```

Comment:
It would be useful to name the methods here (e.g. FrodoMatrix::mul_add_sb_plus_e) and then we can refer back to this location when the notes/bullet points for the KeyGen, Encaps and Decaps refer to 'a specialised function' later on.

Reply:
Similar discussion as https://github.com/sehlen-bsi/botan-docs/pull/117#discussion_r1480040327 - in previous cryptodocs, the text became very unreadable due to a lot of long C++ function names, so we decided against that. But we could make the modification if you see it as important in this case.